tracker/resource.go

```diff
@@ -0,0 +1,179 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tracker
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/apis/duck"
+	duckv1alpha1 "knative.dev/pkg/apis/duck/v1alpha1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/injection/clients/dynamicclient"
+)
+
+// ResourceInformer is an informer that allows tracking arbitrary Resources.
+//
+// The intended use is creating a Reconciler:
+//     resourceInformer := duck.NewResourceInformer(ctx)
+//     r := &Reconciler{ ... }
+//     impl := controller.NewImpl(r, r.Logger, ReconcilerName)
+//     // resourceTracker is used to notify this reconciler when a tracked resource changes.
+//     resourceTracker: resourceInformer.NewTracker(impl.EnqueueKey, controller.GetTrackerLease(ctx))
+type ResourceInformer interface {
+	NewTracker(callback func(string), lease time.Duration) ResourceTracker
+}
+
+// ResourceTracker is a tracker capable of tracking arbitrary Resources.
+//
+// The intended use is inside a reconciler's reconcile loop:
+// func (r *Reconciler) reconcile(ctx context.Context, t *v1alpha1.Trigger) error {
+//     ...
+//     // Tell resourceTracker to reconcile this Trigger whenever the Broker changes.
+//     track := r.resourceTracker.TrackInNamespace(t)
+//     if err = track(utils.ObjectRef(broker, brokerGVK)); err != nil {
+//         return err
+//     }
+//     ...
+type ResourceTracker interface {
+	// TrackInNamespace returns a function that can be used to watch arbitrary Resources in the same
+	// namespace as obj. Any change will cause a callback for obj.
+	TrackInNamespace(obj metav1.Object) func(corev1.ObjectReference) error
+}
+
+// resourceInformer is a concrete implementation of ResourceInformer. It caches informers and
+// ensures TypeMeta.
+type resourceInformer struct {
+	duck duck.InformerFactory
+
+	concrete     map[schema.GroupVersionResource]cache.SharedIndexInformer
+	concreteLock sync.RWMutex
+}
+
+// NewResourceInformer creates a new ResourceInformer.
+func NewResourceInformer(ctx context.Context) ResourceInformer {
+	return &resourceInformer{
+		duck: &duck.TypedInformerFactory{
+			Client:       dynamicclient.Get(ctx),
+			Type:         &duckv1alpha1.AnyResource{},
+			ResyncPeriod: controller.GetResyncPeriod(ctx),
+			StopChannel:  ctx.Done(),
+		},
+		concrete: map[schema.GroupVersionResource]cache.SharedIndexInformer{},
+	}
+}
+
+// NewTracker creates a new ResourceTracker, backed by this ResourceInformer.
+func (i *resourceInformer) NewTracker(callback func(string), lease time.Duration) ResourceTracker {
+	return &resourceTracker{
+		informer: i,
+		tracker:  New(callback, lease),
+		concrete: map[schema.GroupVersionResource]struct{}{},
```

Comment:
Your both usages for `concrete` (here and below) seem to be prime examples for `sync.Map`.
Will make your code much readable.